components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt

```diff
@@ -21,6 +25,63 @@ const val MAX_FAILURE_REASON_LENGTH = 100
  */
 @Suppress("LargeClass")
 object SyncTelemetry {
+    private val logger = Logger("SyncTelemetry")
+
+    /**
+     * Process [SyncTelemetryPing] as returned from [mozilla.appservices.syncmanager.SyncManager].
+     */
+    fun processSyncTelemetry(
+        syncTelemetry: SyncTelemetryPing,
+
+        // The following are present to make this function testable. In tests, we need to "intercept"
+        // values set in singleton ping objects before they're reset by a 'submit' call.
+        submitGlobalPing: () -> Unit = { Pings.sync.submit() },
+        submitHistoryPing: () -> Unit = { Pings.historySync.submit() },
+        submitBookmarksPing: () -> Unit = { Pings.bookmarksSync.submit() },
+        submitLoginsPing: () -> Unit = { Pings.loginsSync.submit() }
+    ) {
+        syncTelemetry.syncs.forEach { syncInfo ->
+            // Note that `syncUuid` is configured to be submitted in all of the sync pings (it's set
+            // once, and will be attached by glean to history-sync, bookmarks-sync, and logins-sync pings).
+            // However, this only happens if sync telemetry is being submitted via [processSyncTelemetry].
+            // That is, if different data types were synchronized together, as happens when using a sync manager.
+            // We can then use 'syncUuid' to associate together all of the individual syncs that happened together.
+            // If a data type sync is executed individually via the legacy 'sync' API on specific storage layers,
+            // then the corresponding ping will not have 'syncUuid' set.
+            val globalSyncUuid = UUID.randomUUID()
+            Sync.syncUuid.set(globalSyncUuid)
+
+            // It's possible for us to sync some engines, and then get a hard error that fails the
+            // entire sync. Examples of such errors are an HTTP server error, token authentication
+            // error, or other kind of network error.
+            // We can have some engines that succeed (and others that fail, with different reasons)
+            // and still have a global failure_reason.
+            syncInfo.failureReason?.let { failureReason ->
+                recordFailureReason(failureReason, Sync.failureReason)
+            }
+
+            syncInfo.engines.forEach { engineInfo ->
+                when (engineInfo.name) {
+                    "bookmarks" -> {
+                        individualBookmarksSync(syncTelemetry.uid, engineInfo, globalSyncUuid)
```

Comment:
I don't think you need to pass `globalSyncUuid` around, see my comments below.

Reply:
Thanks, that makes sense. I guess I was concerned that the value would be cleared after ping submission, but I'm guessing it's cleared for a specific ping that's being submitted, and not for others?
Tests seem to pass with that change.

Reply:
Yes, values are stored in ping-specific stores. This is the big feature that allows easy configuration of new custom ping and adding new metrics to existing ping.

Our docs somethimes _can_ be confusing! Please let us know if they are in this case!